src/Perpetuum.RequestHandlers/ZoneSaveLayer.cs

```diff
@@ -6,24 +6,36 @@ namespace Perpetuum.RequestHandlers
     public class ZoneSaveLayer : IRequestHandler
     {
         private readonly IZoneManager _zoneManager;
-        private readonly LayerFileIO _layerFileIO;
+        private readonly ILayerFileIO _layerFileIO;
 
-        public ZoneSaveLayer(IZoneManager zoneManager,LayerFileIO layerFileIO)
+        public ZoneSaveLayer(IZoneManager zoneManager, ILayerFileIO layerFileIO)
         {
             _zoneManager = zoneManager;
             _layerFileIO = layerFileIO;
         }
 
         public void HandleRequest(IRequest request)
         {
-            foreach (var zone in _zoneManager.Zones)
+            if (request.Data.ContainsKey(k.zoneID))
             {
-                _layerFileIO.SaveLayerToDisk(zone,zone.Terrain.Altitude);
-                _layerFileIO.SaveLayerToDisk(zone,zone.Terrain.Blocks);
-                _layerFileIO.SaveLayerToDisk(zone,zone.Terrain.Controls);
-                _layerFileIO.SaveLayerToDisk(zone,zone.Terrain.Plants);
+                var zoneId = request.Data.GetOrDefault<int>(k.zoneID);
+                _zoneManager.ContainsZone(zoneId).ThrowIfFalse(ErrorCodes.ZoneNotFound);
+                var zone = _zoneManager.GetZone(zoneId);
+                _layerFileIO.SaveLayerToDisk(zone, zone.Terrain.Altitude);
```

Comment:
Move the layer saving group to its own method. DRY 🙂

Reply:
good point